### PR TITLE
Start typing: add mypy to tox and type parser module

### PR DESCRIPTION
**Pull request by pcorpet**

### CODE REVIEW

File: cssselect/parser.py
Comment (wRAR) [inline cssselect/parser.py:175]: Is this intended to be deleted?

Comment (pcorpet) [reply cssselect/parser.py:175]: Yes it is: it was never run: self.selector doesn't exist so it would crash if it was used. `FunctionalPseudoElement` is never used in a tree but separately so it doesn't need to have the same interface as the other classes around.

**Review by wRAR — APPROVED** [review]

**Comment by Gallaecio** [comment]
There are now conflicts due to merging the black pull request

**Comment by pcorpet** [comment]
Conflicts fixed.

**Comment by Gallaecio** [comment]
Note: We are going to keep this on-hold for a few weeks, to avoid conflicts with some on-going work. @elacuesta will review and merge this change once that other work is merged.

**Comment by pcorpet** [comment]
I can add more here. Please tell me when I should type again.

File: cssselect/xpath.py
Comment (wRAR) [inline cssselect/xpath.py:694]: This assert looks wrong as the code below has a `if value` condition

Comment (wRAR) [reply cssselect/xpath.py:694]: Thoug it's not really clear to me how does that code work. I assume `value` can be an empty string but cannot be None? In any case it makes no sense to call xpath_literal() on False or None, and for an empty string the result could be hardcoded instead.

**Comment by wRAR** [comment]
The docs build fails with 

> cssselect/parser.py:docstring of cssselect.parser.FunctionalPseudoElement:1:py:class reference target not found: cssselect.parser.Token

**Comment by Gallaecio** [comment]
> OTOH I'm not sure how can a public method have an argument of a type of a non-public class.

That could be the case if you are supposed to get the argument from some other public API, e.g.

```python
token = public_api.get_token() 
foo = public_api.Foo(token)
```

I am not saying it makes sense here, though.

However, maybe we could leave that for a separate ticket, and solve things here by adding that `autoclass` to some new section about private classes or something? Or even relax the documentation check until we get a proper look at this as part of that separate ticket?

File: cssselect/xpath.py
Comment (wRAR) [reply cssselect/xpath.py:683]: Fun fact: `[attribute!=value]` is a cssselect-specific extension which is not in any CSS standard. This also means its behavior is not described anywhere. So let's look how it works:

```
In [3]: GenericTranslator().css_to_xpath('[foo!=bar]')
Out[3]: "descendant-or-self::*[not(@foo) or @foo != 'bar']"

In [5]: GenericTranslator().css_to_xpath('[foo!=""]')
Out[5]: "descendant-or-self::*[@foo != '']"
```

So valid matches for non-empty values are an absent attribute and an attribute that isn't equal to the value. The only valid match for an empty value is a present non-empty attribute. We should keep this behavior.

This also applies to `xpath_attrib_equals` above, where value can't be None but can be an empty string.

**Pull request merged**
